Integrate Soda Agent with secrets manager (#588)
sodadata/docs: soda-agent/secrets.md

### Prerequisites
* You have set up a Kubernetes cluster in your cloud services environment and deployed a Soda Agent in the cluster.
* You have set up and are using a Hashicorp vault which contains the key-value pairs for your Soda Cloud account username and password: `SODA_USERNAME` and `SODA_PASSWORD`.

### Install and set up the External Secrets Operator
1. Use helm to install the External Secrets Operator from the <a href="https://charts.external-secrets.io" target="_blank">helm chart repository</a> into the same Kubernetes cluster in which you deployed your Soda Agent. 
    ```shell
    helm repo add external-secrets https://charts.external-secrets.io

    helm install external-secrets \
       external-secrets/external-secrets \
        -n external-secrets \
        --create-namespace
    ```
2. Verify the installation using the following command:
```shell
kubectl -n external-secrets get all
```
3. Follow the <a href="https://developer.hashicorp.com/vault/tutorials/kubernetes/kubernetes-sidecar#install-the-vault-helm-chart" target="_blank">Hashicorp Vault instructions</a> to install the Hashicorp Vault helm chart. 

> 4. Follow the <a href="https://developer.hashicorp.com/vault/tutorials/kubernetes/kubernetes-sidecar#set-a-secret-in-vault" target="_blank">Hashicorp Vault instructions</a> to create a secret in which to store your Hashicorp Vault access credentials. The `ClusterSecretStore` uses this secret. This example uses a secret named `external-secrets-vault-app-role-secret-id`.

Reviewer: I personally think that this is to much for this guide and since we wrote at the top in the `Prerequisites` that we assume they have a working Hashicorp Vault instance these lines are not needed.

> 5. Create a `cluster-secret-store.yml` file for the `ClusterSecretStore` configuration. This example uses <a href="https://external-secrets.io/latest/provider/hashicorp-vault/#approle-authentication-example" target="_blank">Hashicorp Vault AppRole authentication</a>. AppRole authenticates with the vault using the <a href="https://developer.hashicorp.com/vault/docs/auth/approle" target="_blank">App Role auth mechanism</a> to access the contents of the vault. It uses the `secretRef` and `roleID` to acquire a temporary access token so that it can fetch secrets.

Reviewer: ```suggestion
5. Create a `cluster-secret-store.yml` file for the `ClusterSecretStore` configuration. This example uses <a href="https://external-secrets.io/latest/provider/hashicorp-vault/#approle-authentication-example" target="_blank">Hashicorp Vault AppRole authentication</a>. AppRole authenticates with Vault using the <a href="https://developer.hashicorp.com/vault/docs/auth/approle" target="_blank">App Role auth mechanism</a> to access the contents of the secret store. It uses the SecretID in the Kubernetes secret referenced by `secretRef` and the `roleID` to acquire a temporary access token so that it can fetch secrets.
```

                key: appRoleSecretId
                name: external-secrets-vault-app-role-secret-id
                namespace: external-secrets
          path: kv
          server: http://vault.vault.svc.cluster.local:8200
          version: v2
    ```
6. Deploy the `ClusterSecretStore` to your cluster.
```shell
kubectl apply -f cluster-secret-store.yaml
```
7. Create an `soda-secret.yml` file for the `ExternalSecret` configuration.
```yaml
apiVersion: external-secrets.io/v1beta1
kind: ExternalSecret
metadata:
  name: soda-agent
  namespace: soda-agent
spec:
  data:
  - remoteRef:
          key: local/soda
          property: SODA_USERNAME
        secretKey: SODA_USERNAME
  - remoteRef:
          key: local/soda
          property: SODA_PASSWORD
        secretKey: SODA_PASSWORD
  refreshInterval: 1m
  secretStoreRef:
        kind: ClusterSecretStore
        name: vault-app-role
  target:
        name: soda-agent-secrets
        template:
          data:
            soda-agent.conf: |
              SODA_USERNAME={% raw %}{{ .SODA_USERNAME }}{% endraw %}
              SODA_PASSWORD={% raw %}{{ .SODA_PASSWORD }}{% endraw %}
          engineVersion: v2
```
This example identifies:

> * the `namespace` of the cluster

Reviewer: ```suggestion
* the `namespace` of the Soda Agent
```

> * two `remoteRef` configurations, including the filepath in the vault, one each for the Soda Cloud account username and password, to detail what the `ExternalSecret` must fetch from the Hashicorp Vault

Reviewer: ```suggestion
* two `remoteRef` configurations, including the path in the vault, one each for `SODA_USERNAME` and `SODA_PASSWORD`, to detail what the `ExternalSecret` must fetch from the Hashicorp Vault
```

* a `refreshInterval` to indicate how often the ESO must reconcile the `remoteRef` values
* the `secretStoreRef` to indicate the `ClusterSecretStore` through which to access the vault

> * a `target template` that creates a file called `soda-agent.conf` into which it adds the username and password values in the `.env` format that the Soda Agent expects.

Reviewer: ```suggestion
* a `target template` that creates a file called `soda-agent.conf` into which it adds the username and password values in the [dotenv](https://saurabh-kumar.com/python-dotenv/#file-format) format that the Soda Agent expects.
```

8. Deploy the `ExternalSecret` to your cluster.
```shell

> kubectl -n external-secrets apply -f soda-secret.yaml

Reviewer: ```suggestion
kubectl  apply -n soda-agent -f soda-secret.yaml
```

Reviewer: Since the namespace is already set in the Kubernetes manifest the `-n` is actually not needed but I think we can leave it for completeness.

```

> 9. Use the following command to get the `ExternalSecret` to authenticate to the Hashicorp Vault using the `ClusterSecretStore`.

Reviewer: Are you trying to get the secret with which ESO connects to Vault or do you want to see the K8S secret which was created by the ExternalSecret ? If the later then the command should be:
```
kubectl get secret -n soda-agent soda-agent-secrets
```

Author: It was the latter, so I'll correct that!
